File: dev/ideas/cppgen/stateupdater.cpp
#include "brianlib.h"
#include<sstream>

void LinearStateUpdater::__call__(NeuronGroup *group)
{
//    n = len(P)
//    m = len(self)
//    S = P._S
//    A = self.A
//    c = self._C
	int m = this->M_n;
	int n = group->num_neurons;
	double *A = this->M;
	double *c = this->b;
    double x[m];
    //for(int i=0;i<m;i++) cout << c[i] << endl;
    for(int i=0;i<n;i++)  
    {
        for(int j=0;j<m;j++)
        {
            x[j] = c[j];
            for(int k=0;k<m;k++)
                //x[j] += A(j,k) * S(k,i);
            	//x[j] += A[j+k*m] * S[k+i*m];
            	//x[j] += A[k+j*m] * S[i+k*n];
            	x[j] += A[k+j*m] * neuron_value(group, i, k);
        }
        for(int j=0;j<m;j++)
            //S[j+i*m] = x[j];
        	//S[i+j*n] = x[j];
        	neuron_value(group, i, j) = x[j];
    }	
}

string LinearStateUpdater::__repr__()
{
	int m = this->M_n;
	double *A = this->M;
	double *c = this->b;
	stringstream out;
	out << "LinearStateUpdater" << endl;
	out << "A = " << endl;
	for(int i=0;i<m;i++)
	{
		for(int j=0;j<m;j++)
			out << A[j+i*m] << " ";
		out << endl;
	}
	out << "c = ";
	for(int i=0;i<m;i++)
		out << c[i] << " ";
	return out.str();	
}

File: brian/experimental/codegen2/stateupdater.py
from brian import *
from integration import *
from symbols import *
from blocks import *
from resolution import *

__all__ = ['CodeGenStateUpdater']

class CodeGenStateUpdater(StateUpdater):
    '''
    State updater using code generation, supports Python, C++, GPU.
    
    Initialised with:
    
    ``group``
        The :class:`~brian.NeuronGroup` that this will be used in.
    ``method``
        The integration method, currently one of :func:`euler`, :func:`rk2`
        or :func:`exp_euler`, but you can define your own too. See
        :func:`make_integration_step` for details.
    ``language``
        The :class:`Language` object.
        
    Creates a :class:`Block` from the equations and the ``method``, gets a set
    of :class:`Symbol` objects from :func:`get_neuron_group_symbols`, and
    defines the symbol ``_neuron_index`` as a :class:`SliceIndex`. Then calls
    :meth:`CodeItem.generate` to get the :class:`Code` object.
    
    Inserts ``t`` and ``dt`` into the namespace, and ``_num_neurons`` and
    ``_num_gpu_indices`` in case they are needed.
    '''
    def __init__(self, group, method, language, clock=None):
        self.clock = guess_clock(clock)
        self.group = group
        eqs = group._eqs
        self.eqs = eqs
        self.method = method
        self.language = language
        block = Block(*make_integration_step(self.method, self.eqs))
        symbols = get_neuron_group_symbols(group, self.language)
        symbols['_neuron_index'] = SliceIndex('_neuron_index',
                                              '0',
                                              '_num_neurons',
                                              self.language,
                                              all=True)
        self.code = block.generate('stateupdate', self.language, symbols)
        log_info('brian.codegen2.CodeGenStateUpdater', 'STATE UPDATE CODE:\n'+self.code.code_str)
        log_info('brian.codegen2.CodeGenStateUpdater', 'STATE UPDATE NAMESPACE KEYS:\n'+str(self.code.namespace.keys()))
        ns = self.code.namespace
        ns['t'] = 1.0 # dummy value
        ns['dt'] = group.clock._dt
        ns['_num_neurons'] = len(group)
        ns['_num_gpu_indices'] = len(group)
    def __call__(self, G):
        code = self.code
        ns = code.namespace
        ns['t'] = G.clock._t
        code()
